### PR TITLE
Allow ELMo to accept higher dimensional input

### PR DESCRIPTION
**Pull request by matt-peters**
@tusharkhot

### CODE REVIEW

**Comment by tusharkhot** [comment]
Thanks! This would help to embed a `ListField[TextField]`. I tested it out on my code and you would also need to test for `j>2` here:
```
    for i, j in enumerate(sequence_lengths):
        tensor_without_boundary_tokens.data[i, :(j - 2), :] = tensor.data[i, 1:(j - 1), :]
        new_mask[i, :(j - 2)] = 1
```
https://github.com/allenai/allennlp/blob/master/allennlp/nn/util.py#L832-L834

Normally you would never have empty strings but a `ListField[TextField]` would have padded empty TextFields. It seems boundary sentence ids are still getting added to these empty text fields and so they still contain the two begin and end boundary tokens i.e. j = 2 in the snippet above.

**Comment by matt-peters** [comment]
Thanks, that's a corner case I hadn't tested.  Let me know if the latest commit doesn't fix it for you.

File: tests/modules/elmo_test.py [hunk 1]
```diff
@@ -131,6 +133,18 @@ def test_elmo(self):
         assert list(elmo_representations[1].size()) == [2, 7, 32]
         assert list(mask.size()) == [2, 7]
 
+    def test_elmo_4D_input(self):
```
Comment (DeNeutoy) [inline tests/modules/elmo_test.py]: Can you change this test to use an expanded input with a size other than 1? That would be more robust.

An even better test would be to check that spitting up the tensor into 3D tensors and running it through elmo results in the same thing as the 4D version.

File: tests/modules/elmo_test.py [hunk 2]
```diff
@@ -131,6 +133,18 @@ def test_elmo(self):
...
+        # Add an extra dimension to character_ids
+        character_ids = self.character_ids.unsqueeze(1)
+        print(character_ids.size())
```
Comment (DeNeutoy) [inline tests/modules/elmo_test.py]: Remove print statement.

File: tests/modules/elmo_test.py [hunk 3]
```diff
@@ -131,6 +133,18 @@ def test_elmo(self):
...
+
+        output = self.elmo(character_ids)
+        elmo_representations = output['elmo_representations']
+        mask = output['mask']
+
+        assert list(elmo_representations[0].size()) == [2, 1, 7, 32]
+        assert list(elmo_representations[1].size()) == [2, 1, 7, 32]
+        assert list(mask.size()) == [2, 1, 7]
```
Comment (DeNeutoy) [inline tests/modules/elmo_test.py]: For the mask, you can actually check that the elements are what you expect, rather than the shape. Obviously this is harder for the tensors, so no worries about that.

File: tests/nn/util_test.py [hunk 4]
```diff
@@ -595,19 +595,18 @@ def test_add_sentence_boundary_token_ids_handles_3D_input(self):
     def test_remove_sentence_boundaries(self):
         tensor = Variable(torch.from_numpy(numpy.random.rand(3, 5, 7)))
         mask = Variable(torch.from_numpy(
-                numpy.array([[1, 1, 1, 0, 0],
```
Comment (DeNeutoy) [inline tests/nn/util_test.py]: What does this change do?

Comment (DeNeutoy) [reply tests/nn/util_test.py]: Oh is it testing that it works if you have only `<S>` and `</S>` tokens? If so, can you add that as a comment.

Comment (matt-peters) [reply tests/nn/util_test.py]: Yes, that's it.  Added a comment.

File: tests/modules/elmo_test.py (hunk 1 above)
Comment (matt-peters) [reply tests/modules/elmo_test.py]: Updated the test to run both 3D and 4D versions and compare.
